### DISCUSSION TITLE
TextDisplay scrollbars

### DISCUSSION THREAD

**Discussion by Booplicate** [fltk-rs/fltk-rs#1273] [Q&A] [answered]
Hello, I couldn't find a relevant issue, so opened this discussion.

I have a `TextDisplay` which contains a lot of text, I want to let the user to scroll it, but I also want to apply my custom styles to it. I can't figure out how to change the scrollbars. I know I can change align and size, but I need some deep customisation (at least change its colour).

I could set a custom `draw` method, but then I'd need to redo all the code for text render, wrapping, and scrolling too; I'm not sure if I can handle that just yet.
I could also draw my own scrollbars which I can change, but I don't know how to hide the existing scrollbars.

**Comment by MoAlyousef** [comment, marked as answer]
Hello
FLTK doesn't expose the TextDisplay's scrollbars unfortunately. You can hide them by passing a size -1 to the set_scrollbar_size method.

**Reply by Booplicate** [reply]
That's what I was afraid of, I also couldn't find anything in FLTK docs. I'll try hiding them and rendering my own.

**Reply by Booplicate** [reply]
Hmm, I can't wrap my head around scrollbars.
Since `TextBuffer.scroll` takes a line to scroll to, I need to know how many lines are available. I can get all the lines via `TextBuffer.count_lines`, but I need to know how many are scrollable (not visible).
For example if I have 15 lines, 5 visible, 10 not visible, I use 10-1 as my maximum scroll value (`Scrollbar.set_maximum`) and 0 as minimum. But I don't have that information from the code. Am I missing something?

**Reply by MoAlyousef** [reply]
I tried to implement a quick example here:
```rust
use fltk::{prelude::*, *, enums::*};

fn main() {
    
    let app = app::App::default();
    let mut buf = text::TextBuffer::default();
    let s = std::fs::read_to_string("CODE_OF_CONDUCT.md").unwrap();
    buf.set_text(&s);

    let mut win = window::Window::default().with_size(400, 300);
    let mut disp = text::TextDisplay::new(5,5,380, 290, None);
    disp.set_buffer(buf.clone());
    disp.set_scrollbar_size(-1);
    let mut slider = valuator::NiceSlider::new(385, 5, 10, 290, None);
    let lc = disp.count_lines(0, buf.length() - 400, true);
    dbg!(lc);
    slider.set_maximum(lc as f64);
    win.end();
    win.show();

    disp.handle({
        let mut s = slider.clone();
        let mut curr = 0;
        move |d, ev| {
        match ev {
            Event::MouseWheel => {
                if app::event_dy() == app::MouseWheel::Up && curr < lc - 3 { // the 3 is because each mousewheel scrolls 3 lines
                    curr += 3;
                } else if app::event_dy() == app::MouseWheel::Down && curr > 0 {
                    curr -= 3;
                }
                s.set_value(curr as _);
                dbg!(curr);
                true
            }
            _ => false,
        }
    }});

    slider.set_callback(move|s| {
        disp.scroll(s.value() as _, 0);
    });

    app.run().unwrap();
}
```

I have a curr value which counts the mousewheel events for the text display, and passes it to the slider. I set the slider's callback to scroll the text display.

**Reply by Booplicate** [reply]
Big thanks for a good example, it helped me *a lot* to get started!
I played around for a bit and figured that it's better to get rid of `curr` and just fetch current value from the slider itself (`slider.value()`), otherwise I had to implement `handle` for the slider which got messy.

I have one more question, why did you do `buf.length() - 400`? I notice if I remove it, my slider becomes a bit finicky, but I'm concerned if that depends on the buffer length, because my buffer can be both small and big depending on the received data.

**Comment by MoAlyousef** [comment]
You’re welcome. 
`buf.length() - 400` was a bit arbitrary. Like you, I noticed that without it the slider would reach the end of the buffer then still have some ways to go without actually scrolling the display. I guess it’s because of the length of the last line, without wanting to concretely measure it, I went with 400. This could cause issues if the length of the buffer is less than 400, so it needs to be more concrete.
